source/module_elecstate/module_charge/charge.cpp

```diff
@@ -342,901 +215,485 @@ void Charge::renormalize_rho(void)
-		ModuleBase::GlobalFunc::ZEROS(rhog[is], this->rhopw->npw);
-		ModuleBase::GlobalFunc::ZEROS(rho_save[is], this->rhopw->nrxx);
-		ModuleBase::GlobalFunc::ZEROS(rhog_save[is], this->rhopw->npw);
-	}
-
-    ModuleBase::Memory::record("Chg::rho", sizeof(double) * GlobalV::NSPIN*this->rhopw->nrxx);
-    ModuleBase::Memory::record("Chg::rho_save", sizeof(double) * GlobalV::NSPIN*this->rhopw->nrxx);
-    ModuleBase::Memory::record("Chg::rhog", sizeof(double) * GlobalV::NSPIN*this->rhopw->npw);
-    ModuleBase::Memory::record("Chg::rhog_save", sizeof(double) * GlobalV::NSPIN*this->rhopw->npw);
+    // allocate memory
+    rho = new double*[GlobalV::NSPIN];
+    rhog = new std::complex<double>*[GlobalV::NSPIN];
+    rho_save = new double*[GlobalV::NSPIN];
+    rhog_save = new std::complex<double>*[GlobalV::NSPIN];
+
+    for (int is = 0; is < GlobalV::NSPIN; is++)
+    {
+        rho[is] = new double[this->rhopw->nrxx];
+        rhog[is] = new std::complex<double>[this->rhopw->npw];
+        rho_save[is] = new double[this->rhopw->nrxx];
+        rhog_save[is] = new std::complex<double>[this->rhopw->npw];
+        ModuleBase::GlobalFunc::ZEROS(rho[is], this->rhopw->nrxx);
+        ModuleBase::GlobalFunc::ZEROS(rhog[is], this->rhopw->npw);
+        ModuleBase::GlobalFunc::ZEROS(rho_save[is], this->rhopw->nrxx);
+        ModuleBase::GlobalFunc::ZEROS(rhog_save[is], this->rhopw->npw);
+    }
+
+    ModuleBase::Memory::record("Chg::rho", sizeof(double) * GlobalV::NSPIN * this->rhopw->nrxx);
+    ModuleBase::Memory::record("Chg::rho_save", sizeof(double) * GlobalV::NSPIN * this->rhopw->nrxx);
+    ModuleBase::Memory::record("Chg::rhog", sizeof(double) * GlobalV::NSPIN * this->rhopw->npw);
+    ModuleBase::Memory::record("Chg::rhog_save", sizeof(double) * GlobalV::NSPIN * this->rhopw->npw);
 
     this->rho_core = new double[this->rhopw->nrxx]; // core charge in real space
-    ModuleBase::GlobalFunc::ZEROS( rho_core, this->rhopw->nrxx);
+    ModuleBase::GlobalFunc::ZEROS(rho_core, this->rhopw->nrxx);
 
-	this->rhog_core = new std::complex<double>[this->rhopw->npw]; // reciprocal core charge
-	ModuleBase::GlobalFunc::ZEROS( rhog_core, this->rhopw->npw);
+    this->rhog_core = new std::complex<double>[this->rhopw->npw]; // reciprocal core charge
+    ModuleBase::GlobalFunc::ZEROS(rhog_core, this->rhopw->npw);
 
     ModuleBase::Memory::record("Chg::rho_core", sizeof(double) * this->rhopw->nrxx);
     ModuleBase::Memory::record("Chg::rhog_core", sizeof(double) * this->rhopw->npw);
 
-	this->allocate_rho_final_scf = true;
+    this->allocate_rho_final_scf = true;
     return;
 }
 
 //=========================================================
 // calculate total number of electrons (GlobalV::nelec) and default
 // number of bands (GlobalV::NBANDS).
 //=========================================================
-#include "module_elecstate/occupy.h"
-void Charge::cal_nelec(void)
+//#include "module_elecstate/occupy.h"
+void Charge::cal_nelec(const UnitCell& ucell)
```

Comment:
It seems function "cal_nelec()" should belong to UnitCell but not Charge.

Reply:
Thanks, maybe you are right. But here this function claculates ne from rho, and cell does not have.the information of rho. How to solve this conflict?@dyzheng

Reply:
@dyzheng I will do this refactor in a separate PR.